include/kernels/restrictor.cuh

```diff
@@ -2,211 +2,166 @@
+      aggregate_size(in.Volume()/out.Volume()),
+      aggregate_size_cb(in.VolumeCB()/out.Volume()),
+      fine_to_coarse(fine_to_coarse), coarse_to_fine(coarse_to_fine),
+      spin_map(), parity(parity), nParity(in.SiteSubset()), swizzle_factor(1)
     { }
   };
 
   /**
      Rotates from the fine-color basis into the coarse-color basis.
   */
-  template <typename Float, int fineSpin, int fineColor, int coarseColor, int coarse_colors_per_thread,
-	    class FineColor, class Rotator>
-  __device__ __host__ inline void rotateCoarseColor(complex<Float> out[fineSpin*coarse_colors_per_thread],
-						    const FineColor &in, const Rotator &V,
-						    int parity, int nParity, int x_cb, int coarse_color_block) {
-    const int spinor_parity = (nParity == 2) ? parity : 0;
-    const int v_parity = (V.Nparity() == 2) ? parity : 0;
+  template <typename Out, typename Arg>
+  __device__ __host__ inline void rotateCoarseColor(Out &out, const Arg &arg, int parity, int x_cb, int coarse_color_block)
+  {
+    constexpr int coarse_color_per_thread = coarse_colors_per_thread<Arg::fineColor, Arg::coarseColor>();
+    const int spinor_parity = (arg.nParity == 2) ? parity : 0;
+    const int v_parity = (arg.V.Nparity() == 2) ? parity : 0;
 
 #pragma unroll
-    for (int s=0; s<fineSpin; s++)
+    for (int s=0; s<Arg::fineSpin; s++)
 #pragma unroll
-      for (int coarse_color_local=0; coarse_color_local<coarse_colors_per_thread; coarse_color_local++) {
-        out[s*coarse_colors_per_thread+coarse_color_local] = 0.0;
+      for (int coarse_color_local=0; coarse_color_local<coarse_color_per_thread; coarse_color_local++) {
+	out[s*coarse_color_per_thread+coarse_color_local] = 0.0;
       }
 
 #pragma unroll
-    for (int coarse_color_local=0; coarse_color_local<coarse_colors_per_thread; coarse_color_local++) {
+    for (int coarse_color_local=0; coarse_color_local<coarse_color_per_thread; coarse_color_local++) {
       int i = coarse_color_block + coarse_color_local;
 #pragma unroll
-      for (int s=0; s<fineSpin; s++) {
+      for (int s=0; s<Arg::fineSpin; s++) {
 
-	constexpr int color_unroll = fineColor == 3 ? 3 : 2;
+	constexpr int color_unroll = Arg::fineColor == 3 ? 3 : 2;
 
-	complex<Float> partial[color_unroll];
+	complex<typename Arg::real> partial[color_unroll];
 #pragma unroll
 	for (int k=0; k<color_unroll; k++) partial[k] = 0.0;
 
 #pragma unroll
-	for (int j=0; j<fineColor; j+=color_unroll) {
+	for (int j=0; j<Arg::fineColor; j+=color_unroll) {
 #pragma unroll
 	  for (int k=0; k<color_unroll; k++)
-	    partial[k] += conj(V(v_parity, x_cb, s, j+k, i)) * in(spinor_parity, x_cb, s, j+k);
+	    partial[k] = cmac(conj(arg.V(v_parity, x_cb, s, j+k, i)), arg.in(spinor_parity, x_cb, s, j+k), partial[k]);
 	}
 
 #pragma unroll
-	for (int k=0; k<color_unroll; k++) out[s*coarse_colors_per_thread + coarse_color_local] += partial[k];
+	for (int k=0; k<color_unroll; k++) out[s*coarse_color_per_thread + coarse_color_local] += partial[k];
```

Comment:
`cmac`

Reply:
`cmac` isn't applicable here, there's no multiply here, it's just summation.